SCMOD-12257: Correlation Id Support

simranjeetc opened the pull request:
https://portal.digitalsafe.net/browse/SCMOD-12257

Commit:
Added support correlationId

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=WorkerFramework&REPO_NAME=worker-framework&BRANCH_NAME=SCMOD-12346) to go ahead and create the build...

simranjeetc commented:
@dermot-hardy I found this https://github.com/WorkerFramework/worker-framework/blob/41515328a3060d0499dc764fae07633e9e5f3a3c/worker-core/src/main/java/com/hpe/caf/worker/core/WorkerApplication.java#L192 code where we are checking for presence of logback.xml after which we switch the logging framework to slf4j which can leverage MDC for correlation id. If caf-correaltion-id dependency is added to this project, then logback.xml will always be present which will result in slf4j to be used. This would make if check redundant. Do we need to remove the if block? Or am I missing something?

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/WorkerFramework/job/WorkerFramework~worker-framework~SCMOD-12346~CI

buildmachine-sou-jenkins2 commented:
The Documentation QA site for this branch has been built:
https://pages.github.houston.softwaregrp.net/sepg-docs-qa/WorkerFramework_CI_worker-framework_SCMOD-12346

Commit:
Moved the removal from MDC to after catch block to handle failure scenarios

dermot-hardy commented on worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java:
```diff
@@ -87,6 +87,11 @@
+    /**
+     * This field contains the correlation id which is eventually logged via MDC.
+     */
+    private String correlationID;
```
- Change to `correlationId`, to be consistent with `taskId`.
- Do we need to increment the `TASK_MESSAGE_VERSION`?  (I assume not but I'm not totally sure.)

simranjeetc replied:
Addressed `correlationId` casing. 
`TASK_MESSAGE_VERSION` - I found these instances of where it is getting used http://16.103.38.246:9090/search?group=1_verity&group=2_caf&group=3_Other-FAS-Libraries&full=%22workertask.getVersion%22&defs=&refs=&path=&hist=&type=&xrd=&nn=3&si=full&si=full

In case, we want to reject any messages which do not have this change, we can reject those messages by incrementing `TASK_MESSAGE_VERSION`. For correlation id change, I don't see a need to increment the version. But not entirely sure about any other use-case.

dermot-hardy commented on worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java:
```diff
@@ -107,11 +112,12 @@ public TaskMessage(final String taskId, final String taskClassifier, final int t
     public TaskMessage(final String taskId, final String taskClassifier, final int taskApiVersion, final byte[] taskData,
-                       final TaskStatus taskStatus, final Map<String, byte[]> context, final String to, final TrackingInfo tracking, final TaskSourceInfo sourceInfo)
+                       final TaskStatus taskStatus, final Map<String, byte[]> context, final String to, final TrackingInfo tracking, final TaskSourceInfo sourceInfo,
+                       final String correlationID)
```
Removing the `(String, String, int, byte[], TaskStatus, Map, String, TrackingInfo, TaskSourceInfo)` constructor is a breaking change.

Create an overload instead.

Also update the new parameter name to `correlationId`.

simranjeetc replied:
Done

dermot-hardy commented on worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java:
```diff
@@ -233,4 +240,14 @@ public void setPriority(Integer priority)
         this.priority = priority;
     }
+    
+    public String getCorrelationID()
```
```suggestion
    public String getCorrelationId()
```

simranjeetc replied:
Done

dermot-hardy commented on worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java:
```diff
@@ -233,4 +240,14 @@ public void setPriority(Integer priority)
+        return correlationID;
+    }
+    
+    public void setCorrelationID(String correlationID)
```
```suggestion
    public void setCorrelationId(String correlationId)
```

simranjeetc replied:
Done

dermot-hardy commented on worker-api/src/main/java/com/hpe/caf/api/worker/WorkerTaskData.java:
```diff
@@ -44,6 +44,11 @@
      * Retrieves any task specific context associated with the task
      */
     byte[] getContext();
+    
```
```suggestion

```

dermot-hardy commented on worker-api/src/main/java/com/hpe/caf/api/worker/WorkerTaskData.java:
```diff
@@ -44,6 +44,11 @@
+    /**
+     * Retrieves task specific correlationID
+     */
+    String getCorrelationID();
```
```suggestion
    String getCorrelationId();
```

simranjeetc replied:
Done

dermot-hardy commented on worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java:
```diff
@@ -123,6 +123,12 @@ public TaskStatus getStatus()
     {
         return taskMessage.getContext().get(servicePath.toString());
     }
+    
```
```suggestion

```

dermot-hardy commented on worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java:
```diff
@@ -24,13 +24,15 @@
  */
 class StreamingWorkerWrapper implements Runnable
 {
+    public static final String CORRELATION_ID = "correlationId";
```
Why is this public?

simranjeetc replied:
Changed to private

dermot-hardy commented on worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java:
```diff
@@ -73,6 +76,7 @@ public void run()
             LOG.warn("Worker threw unhandled exception", e);
             workerTask.setResponse(worker.getGeneralFailureResult(e));
         }
+        MDC.remove(CORRELATION_ID);
```
Put this in a `finally` block to make sure that it is done (even if one of the exception handlers throws an exception).

simranjeetc replied:
Done

simranjeetc committed:
Update worker-api/src/main/java/com/hpe/caf/api/worker/WorkerTaskData.java

Co-authored-by: Dermot Hardy <dermot.hardy@microfocus.com>

simranjeetc committed:
Update worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java

Co-authored-by: Dermot Hardy <dermot.hardy@microfocus.com>

Commit:
Addressed review comments -
1. Changed correlationID -> correlationId
2. Added overloaded constructor
3. Removing MDC key in finally block

Commit:
Merge remote-tracking branch 'origin/SCMOD-12346' into SCMOD-12346

dermot-hardy committed:
Removed an unnecessary change

dermot-hardy committed:
Minor whitespace changes
